### PR TITLE
Add calculation of sequencing metrics for bcl2fastq data

### PR DESCRIPTION
**Pull request by seallard**
## Description
This PR adds functionality to the cg sequencing metrics module for calculating metrics from bcl2fastq data.
Please take an extra close look at the functions in [the calculation module](https://github.com/Clinical-Genomics/cg/pull/2045/files#diff-65588aa189f004f5f4e42fb579eadfb08e4649ef3fd6bb48b599be488e90ea42) and let me know if you think they make sense and whether the docs can be improved.

### Added
- Module for converting parsed bcl2fastq metrics to SequencingStatistics
- Module for calculating sequencing metrics
- Tests for calculation of metrics

### This [version](https://semver.org/) is a
- [ ] **MINOR** - when you add functionality in a backwards compatible manner

### CODE REVIEW

File: cg/apps/sequencing_metrics_parser/parsers/bcl2fastq_to_sequencing_statistics.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        sample_internal_id=demux_result.sample_id,
+        lane=conversion_result.lane_number,
+        yield_in_megabases=yield_in_megabases,
+        read_counts=demux_result.number_reads,
+        passed_filter_percent=passed_filter_percent,
+        raw_clusters_per_lane_percent=avg_clusters_per_lane,
+        perfect_index_reads_percent=perfect_reads_ratio,
+        bases_with_q30_percent=bases_with_q30_percent,
+        lanes_mean_quality_score=average_quality_score,
+        started_at=datetime.now(),
```
Comment (seallard) [inline cg/apps/sequencing_metrics_parser/parsers/bcl2fastq_to_sequencing_statistics.py:102]: I'm not really sure what this `started_at`field refers to. When the demultiplexing was started? We should rename it, but in the meantime - what is it? 😅

File: cg/apps/sequencing_metrics_parser/parsers/bcl2fastq_to_sequencing_statistics.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        total_clusters=conversion_result.total_clusters_raw,
+        lane_count=number_of_lanes,
+    )
+
+    return SequencingStatistics(
+        flow_cell_name=raw_sequencing_metrics.flowcell,
+        sample_internal_id=demux_result.sample_id,
+        lane=conversion_result.lane_number,
+        yield_in_megabases=yield_in_megabases,
+        read_counts=demux_result.number_reads,
```
Comment (seallard) [inline cg/apps/sequencing_metrics_parser/parsers/bcl2fastq_to_sequencing_statistics.py:96]: I think a better name for this field is `sample_read_count` or `lane_sample_read_count`. What do you think?

Comment (ChrOertlin) [reply cg/apps/sequencing_metrics_parser/parsers/bcl2fastq_to_sequencing_statistics.py:96]: Yes, now that I know what these field really are it indeed is ```total read count```

**Review by ChrOertlin — APPROVED** [review]
Looks good. 

For me it will be a bit different since some of the metric files have to be summarised per lane and some per read pair and lane... 

Maybe it is a better idea to tie the caluclations to a bcl2fastq class object and work with that.

File: cg/apps/sequencing_metrics_parser/parsers/bcl2fastq_to_sequencing_statistics.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    perfect_reads_ratio: float = calculate_perfect_reads_ratio(demux_result=demux_result)
+
+    number_of_lanes: int = len(raw_sequencing_metrics.conversion_results)
+    avg_clusters_per_lane = average_clusters_per_lane(
+        total_clusters=conversion_result.total_clusters_raw,
+        lane_count=number_of_lanes,
+    )
+
+    return SequencingStatistics(
+        flow_cell_name=raw_sequencing_metrics.flowcell,
```
Comment (ChrOertlin) [inline cg/apps/sequencing_metrics_parser/parsers/bcl2fastq_to_sequencing_statistics.py:92]: maybe you can condense this function that the fields get written to directly with the functions

File: tests/apps/sequencing_metrics_parser/parsers/test_bcl2fastq_to_sequencing_statistics.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    calculate_q30_ratio,
+    calculate_yield_in_megabases,
+)
+from cg.apps.sequencing_metrics_parser.models.bcl2fastq_metrics import ConversionResult, DemuxResult
+
+
+def test_calculate_pass_ratio_from_conversion_result(
+    conversion_result: ConversionResult,
+):
+    # GIVEN a conversion result with valid total clusters
```
Comment (henrikstranneheim) [inline tests/apps/sequencing_metrics_parser/parsers/test_bcl2fastq_to_sequencing_statistics.py:15]:
```suggestion
    # GIVEN a conversion result with valid total clusters

```

File: tests/apps/sequencing_metrics_parser/parsers/test_bcl2fastq_to_sequencing_statistics.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        conversion_result.total_clusters_pf / conversion_result.total_clusters_raw
+    )
+
+    assert pass_filter_ratio == correct_filter_ratio
+
+
+def test_calculate_yield_in_megabases_from_conversion_result(conversion_result: ConversionResult):
+    # GIVEN a conversion result with a known yield
+    for demux_result in conversion_result.demux_results:
+        demux_result.yield_ = 1000000
```
Comment (henrikstranneheim) [inline tests/apps/sequencing_metrics_parser/parsers/test_bcl2fastq_to_sequencing_statistics.py:30]: Use constant

File: tests/apps/sequencing_metrics_parser/parsers/test_bcl2fastq_to_sequencing_statistics.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+    # THEN the pass filter ratio should be the the clusters passing the filter divided by the total clusters
+    correct_filter_ratio: float = (
+        conversion_result.total_clusters_pf / conversion_result.total_clusters_raw
+    )
+
+    assert pass_filter_ratio == correct_filter_ratio
+
+
+def test_calculate_yield_in_megabases_from_conversion_result(conversion_result: ConversionResult):
```
Comment (henrikstranneheim) [inline tests/apps/sequencing_metrics_parser/parsers/test_bcl2fastq_to_sequencing_statistics.py:27]: Add doc str for the tests

File: tests/apps/sequencing_metrics_parser/test_sequencing_metrics_calculator.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    q30_ratio,
+    average_quality_score,
+    yield_in_megabases,
+    pass_filter_ratio,
+    perfect_reads_ratio,
+    average_clusters_per_lane,
+)
+
+
+def test_calculate_yield_in_megabases():
```
Comment (henrikstranneheim) [inline tests/apps/sequencing_metrics_parser/test_sequencing_metrics_calculator.py:11]: Add doc str to tests

File: tests/apps/sequencing_metrics_parser/test_sequencing_metrics_calculator.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    yield_metric: int = yield_in_megabases(total_bases=5000000)
```
Comment (henrikstranneheim) [inline tests/apps/sequencing_metrics_parser/test_sequencing_metrics_calculator.py:12]:
```suggestion
    yield_metric: int = yield_in_megabases(total_bases=5_000_000)
```

**Review by henrikstranneheim — COMMENTED** [review]
Awesome - very clear and easy to read

File: cg/apps/sequencing_metrics_parser/parsers/bcl2fastq_to_sequencing_statistics.py (hunk 1 above)
Comment (seallard) [reply cg/apps/sequencing_metrics_parser/parsers/bcl2fastq_to_sequencing_statistics.py:102]: Ok! Lets do the renaming in a separate PR 👍

File: cg/apps/sequencing_metrics_parser/parsers/bcl2fastq_to_sequencing_statistics.py (hunk 2 above)
Comment (seallard) [reply cg/apps/sequencing_metrics_parser/parsers/bcl2fastq_to_sequencing_statistics.py:96]: Lets do an update of the fields in a separate PR 👍

File: cg/apps/sequencing_metrics_parser/parsers/bcl2fastq_to_sequencing_statistics.py [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+    passed_filter_percent: float = calculate_pass_filter_ratio(conversion_result=conversion_result)
+
+    average_quality_score: float = calculate_average_quality_score(demux_result=demux_result)
+
+    bases_with_q30_percent: float = calculate_q30_ratio(demux_result=demux_result)
+
+    perfect_reads_ratio: float = calculate_perfect_reads_ratio(demux_result=demux_result)
+
+    number_of_lanes: int = len(raw_sequencing_metrics.conversion_results)
+    avg_clusters_per_lane = average_clusters_per_lane(
```
[the remark this answers is not in the document]
Comment (seallard) [reply cg/apps/sequencing_metrics_parser/parsers/bcl2fastq_to_sequencing_statistics.py:86]: I think they are 
```
TotalClustersRaw: The total number of clusters in the lane before filtering.
TotalClustersPF: The total number of clusters in the lane that passed filter (PF).
```

File: cg/apps/sequencing_metrics_parser/parsers/bcl2fastq_to_sequencing_statistics.py (hunk 3 above)
Comment (seallard) [reply cg/apps/sequencing_metrics_parser/parsers/bcl2fastq_to_sequencing_statistics.py:92]: I think that would be confusing, to have the functions mutate the object instead?

File: tests/apps/sequencing_metrics_parser/parsers/test_bcl2fastq_to_sequencing_statistics.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    calculate_perfect_reads_ratio,
+    calculate_q30_ratio,
+    calculate_yield_in_megabases,
+)
+from cg.apps.sequencing_metrics_parser.models.bcl2fastq_metrics import ConversionResult, DemuxResult
+
+
+def test_calculate_pass_ratio_from_conversion_result(
+    conversion_result: ConversionResult,
+):
```
[the remark this answers is not in the document]
Comment (seallard) [reply tests/apps/sequencing_metrics_parser/parsers/test_bcl2fastq_to_sequencing_statistics.py:14]: I think the names of the test functions make them self documenting here, but I added them anyway

File: tests/apps/sequencing_metrics_parser/test_sequencing_metrics_calculator.py (hunk 7 above)
Comment (seallard) [reply tests/apps/sequencing_metrics_parser/test_sequencing_metrics_calculator.py:11]: I do think they are self documenting, but added a doc string anyway to each test

File: cg/apps/sequencing_metrics_parser/parsers/bcl2fastq_to_sequencing_statistics.py [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+    sequencing_statistics: List[SequencingStatistics] = []
+
+    for conversion_result in raw_sequencing_metrics.conversion_results:
+        for demux_result in conversion_result.demux_results:
+            statistics: SequencingStatistics = create_sequencing_statistics(
+                conversion_result=conversion_result,
+                demux_result=demux_result,
+                raw_sequencing_metrics=raw_sequencing_metrics,
+            )
+            sequencing_statistics.append(statistics)
```
[the remark this answers is not in the document]
Comment (seallard) [reply cg/apps/sequencing_metrics_parser/parsers/bcl2fastq_to_sequencing_statistics.py:47]: Renamed it to be a bit more clear to `statistics_for_sample_in_lane`

File: cg/apps/sequencing_metrics_parser/parsers/bcl2fastq_to_sequencing_statistics.py [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+
+    raw_sequencing_metrics: Bcl2FastqSequencingMetrics = parse_bcl2fastq_sequencing_metrics(
+        stats_json_path=stats_json_path
+    )
+
+    sequencing_statistics: List[SequencingStatistics] = []
+
+    for conversion_result in raw_sequencing_metrics.conversion_results:
+        for demux_result in conversion_result.demux_results:
+            statistics: SequencingStatistics = create_sequencing_statistics(
```
[the remark this answers is not in the document]
Comment (seallard) [reply cg/apps/sequencing_metrics_parser/parsers/bcl2fastq_to_sequencing_statistics.py:42]: Renamed the variable to `statistics_for_sample_in_lane`

**Comment by seallard** [comment]
Deployed to stage

**Pull request merged**